security/crypto/bls12/bls12.go

```diff
@@ -350,75 +343,85 @@ func (bls *bls12Base) Combine(signatures ...hotstuff.QuorumSignature) (combined
 			}
 			g2.Add(&agg, &agg, &sig2.sig)
 		} else {
-			bls.logger.Panicf("cannot combine incompatible signature type %T (expected %T)", sig1, sig2)
+			return nil, fmt.Errorf("cannot combine incompatible signature type %T (expected %T)", sig1, sig2)
 		}
 	}
 	return &AggregateSignature{sig: agg, participants: participants}, nil
 }
 
 // Verify verifies the given quorum signature against the message.
-func (bls *bls12Base) Verify(signature hotstuff.QuorumSignature, message []byte) bool {
+func (bls *bls12Base) Verify(signature hotstuff.QuorumSignature, message []byte) (err error) {
 	s, ok := signature.(*AggregateSignature)
 	if !ok {
-		bls.logger.Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
+		return fmt.Errorf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
 	}
 
 	n := s.Participants().Len()
 
 	if n == 1 {
 		id := firstParticipant(s.Participants())
-		pk, ok := bls.publicKey(id)
-		if !ok {
-			bls.logger.Warnf("Missing public key for ID %d", id)
-			return false
+		pk, err := bls.publicKey(id)
+		if err != nil {
+			return err
+		}
+		if !bls.coreVerify(pk, message, &s.sig, domain) {
```

Comment:
If you make `coreVerify` return the error, you can still do it directly like before:
```go
return bls.coreVerify(pk, message, &s.sig, domain)
```